### PR TITLE
6952: Improve Summon Boat spell grayed out messages

### PR DESCRIPTION
**Pull request by Kazhuu**
Solves https://github.com/ihhub/fheroes2/issues/6952 and https://github.com/ihhub/fheroes2/issues/4849 by adding more logic to Summon Boat adventure spell when it is grayed out.

### CODE REVIEW

File: src/fheroes2/heroes/heroes_spell.cpp [hunk 1]
```diff
@@ -303,38 +296,28 @@ namespace
...
+        // Player should have summonable boats before calling this function.
+        assert( !summonableBoats.empty() );
 
-            const uint32_t distance = Maps::GetStraightLineDistance( boatSource, hero.GetIndex() );
+        const int heroColor = hero.GetColor();
 
-            if ( distance > 1 ) {
-                Interface::GameArea & gameArea = Interface::Basic::Get().GetGameArea();
-                gameArea.runSingleObjectAnimation( std::make_shared<Interface::ObjectFadingOutInfo>( tileSource.GetObjectUID(), boatSource, MP2::OBJ_BOAT ) );
+        const int32_t boatSource = summonableBoats.at( 0 );
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_spell.cpp:309]: **summonableBoats.front()** will do the same.

File: src/fheroes2/spell/spell_info.cpp [hunk 2]
```diff
@@ -358,4 +360,32 @@ namespace fheroes2
         }
         return boatDestination;
     }
+
+    Maps::Indexes getSummonableBoats( const Heroes & hero )
```
Comment (ihhub) [inline src/fheroes2/spell/spell_info.cpp:364]: The return array is never used and only first value from it is being used. I suggest to just return an index of a boat. If it is less than 0 then it is an invalid position. In this case the code will be faster.

In this case can we just use **getPossibleBoatPosition()** function for our needs?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Kazhuu , I left multiple comments here. Could you please check them in your spare time?

File: src/fheroes2/spell/spell_info.cpp (hunk 2 above)
Comment (Kazhuu) [reply src/fheroes2/spell/spell_info.cpp]: I changed the function to return possible boat position as int32_t and if it's not found simply return -1 as a result. Reduces code coupling and makes the function simpler to use as well.

**Comment by Kazhuu** [comment]
Hello @ihhub I made the changes. Very good comments. The code should now be simpler as well. Can you review again?

File: src/fheroes2/heroes/heroes_spell.cpp [hunk 3]
```diff
@@ -303,38 +296,27 @@ namespace
...
         const int32_t boatDestination = fheroes2::getPossibleBoatPosition( hero );
         assert( Maps::isValidAbsIndex( boatDestination ) );
 
-        for ( const int32_t boatSource : Maps::GetObjectPositions( center, MP2::OBJ_BOAT, false ) ) {
-            assert( Maps::isValidAbsIndex( boatSource ) );
-
-            Maps::Tiles & tileSource = world.GetTiles( boatSource );
-            const int boatColor = tileSource.getBoatOwnerColor();
-            const int heroColor = hero.GetColor();
+        int32_t boatSource = fheroes2::getSummonableBoat( hero );
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_spell.cpp:302]: This variable can be **const** as we don't modify it after. SonarCloud actually detects this but only for PRs from this main repo.

File: src/fheroes2/spell/spell_info.cpp [hunk 4]
```diff
@@ -348,14 +350,42 @@ namespace fheroes2
...
+    int32_t getSummonableBoat( const Heroes & hero )
+    {
+        const int32_t center = hero.GetIndex();
+
+        for ( const int32_t boatSource : Maps::GetObjectPositions( center, MP2::OBJ_BOAT, false ) ) {
+            assert( Maps::isValidAbsIndex( boatSource ) );
+
+            Maps::Tiles & tileSource = world.GetTiles( boatSource );
+            const int boatColor = tileSource.getBoatOwnerColor();
+            const int heroColor = hero.GetColor();
```
Comment (ihhub) [inline src/fheroes2/spell/spell_info.cpp:372]: This variable can be declared outside the loop. Since it is not inlined we will save some time doing this.

**Comment by Kazhuu** [comment]
Hello @ihhub. Fixed the issues. Can you look again?

**Comment by Branikolog** [comment]
I've made a couple of tests and graying seems to work correctly.
The only thing I'm concerning is wording in a case I summoned a few boats and there're no more available:

https://github.com/ihhub/fheroes2/assets/55348946/5d537cab-0a7b-44fa-9265-560e250875f0

I'm a little bit confused, as the message says, that no boat is available. Actually there're two boats at least available (which I summoned already) but this phrase makes an impression, that no boats can be summoned at all. I believe, this message is actual only for the case, when there're actually no free boats at all which could be summoned.
Maybe, if it would be spelled like: "There's no _more_ boat available to cast this spell.", it would make more clear, that I have available boats, but I simply can't summon more to the current location. How do you think?
Sorry, if I just understood this message incorrectly. :)

**Comment by Kazhuu** [comment]
Good point @Branikolog. I changed to text that you suggested. What you think @LeHerosInconnu?

**Comment by Kazhuu** [comment]
Sorry for the late answer. So what is the text that should be displayed and in what conditions @LeHerosInconnu and @Branikolog? It would be easier for me to fix this if stated clearly :)

**Comment by Branikolog** [comment]
Hi, @Kazhuu !

For the case when there's no any boats available to summon, game should show the message: "There's no boat available to cast this spell."
For the case, when hero has already summoned 1 or more boats, which are standing empty nearby: "There's no more boat available to cast this spell."

**Comment by Kazhuu** [comment]
Okay sounds good @Branikolog but how the text should be reflected on this message "There is no boat available and no free ocean location near the hero to cast this spell."? Should this text also have two variants? To be honest if so I think logic wise this gets quite complicated to implement and the benefit is only little bit better wording for the text.

**Comment by LeHerosInconnu** [comment]
Hello @Kazhuu and @Branikolog,

> Hi, @Kazhuu !
> 
> For the case when there's no any boats available to summon, game should show the message: "There's no boat available to cast this spell." For the case, when hero has already summoned 1 or more boats, which are standing empty nearby: "There's no more boat available to cast this spell."

> Okay sounds good @Branikolog but how the new change should be reflected on this message "There is no boat available and no free ocean location near the hero to cast this spell."? Should this text also have two variants? To be honest if so I think logic wise this gets quite complicated to implement and the benefit is only little bit better wording for the text.

If this becomes too complicated to manage, for me it's okay to have only the sentence "There is no boat available to cast this spell." as it can encompass both cases and is "acceptable" for the description.
@Branikolog, what do you think?

**Comment by Kazhuu** [comment]
I'd say it's not possible to implement, but I myself would like to keep the text more simple :+1:

**Comment by Branikolog** [comment]
@Kazhuu 

Okay then. If it really doesn't worth overcomplicating then I'm OK with initial implementation without adding "more".

I'll probably create an issue for future improvement.

**Review by Branikolog — APPROVED** [review]

**Comment by Kazhuu** [comment]
Thanks @Branikolog! Yeah another issue is good as well. I might work on it after a while, if nobody else takes it :+1:

**Comment by ihhub** [comment]
We will merge this pull request for 1.0.6.

**Comment by ihhub** [comment]
@Kazhuu , thanks a lot for these improvements!

**Pull request merged**